## Walkthrough of Traveler dataset

* **Goal:** Predict the country that users will make their first booking in, based on some basic user profile data.


* **Training data:** set of users with correct category (i.e. what country they made their first booking in).


* **Build a model:** to accurately predict the country of first booking.


* **Test data:** set of users without the knowledge of outcome.


## Walkthrough of Data Scientist process

#### [1] Pre-processing: Assessing and analyzing (Understanding) data, cleaning, transforming and adding new features
#### [2] Learning model: Constructing and testing learning model
#### [3] Post-processing: Creating final predictions


### **First Task in hand:** Preprocessing, how?

### Milestone1: Understanding the Data

#### Formulate range of questions including (but not limited to):

   1. What features (columns) does the dataset contain?
   
   
   2. How many records (rows) have been provided?
   
   
   3. What format is the data in (e.g. what format are the dates provided, are there numerical values, what do the different categorical values look like)?
   
   
   4. Are there missing values?
   
   
   5. How do the different features relate to each other?
    
    Note: Look into the csv files provided.

In [1]:
#from IPython.display import display, Math, Latex


### Reviewing the Dataset

   1. **train_users_2.csv**  – This dataset contains data on Traveler users, including the *destination countries*. Each row represents one user with the columns containing various information such the users’ ages and when they signed up. This is the primary dataset used to train the model.
   
    
   2. **test_users.csv** – This dataset also contains data on Traveler users, in the same format as train_users_2.csv, except without the destination country. These are the users for which final prediction model need to be tested.
   
   
   3. **sessions.csv** – This data is supplementary data that can be used to train the model and make the final predictions. It contains information about the actions (e.g. clicked on a listing, updated a  wish list, ran a search etc.) taken by the users in both the testing and training datasets.



### Exploring Traveler data

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%pylab inline 


Populating the interactive namespace from numpy and matplotlib


In [2]:
#Reading the train_users_2.csv file
data = pd.read_csv("./traveler_dataset/train_users_2.csv")
train_users = data

In [3]:
#Reading the test_users.csv file
test_users = pd.read_csv('./traveler_dataset/test_users.csv')

In [4]:
## How many users are in training set and test set
print("We have", train_users.shape[0], "users in the training set and", 
      test_users.shape[0], "in the test set.")
print("In total we have", train_users.shape[0] + test_users.shape[0], "users.")

We have 213451 users in the training set and 62096 in the test set.
In total we have 275547 users.


In [5]:
train_users.head().transpose()

0               1                2  \
id                           gxn3p5htnn      820tgsjxq7       4ft3gnwmtx   
date_account_created         2010-06-28      2011-05-25       2010-09-28   
timestamp_first_active   20090319043255  20090523174809   20090609231247   
date_first_booking                  NaN             NaN       2010-08-02   
gender                        -unknown-            MALE           FEMALE   
age                                 NaN              38               56   
signup_method                  facebook        facebook            basic   
signup_flow                           0               0                3   
language                             en              en               en   
affiliate_channel                direct             seo           direct   
affiliate_provider               direct          google           direct   
first_affiliate_tracked       untracked       untracked        untracked   
signup_app                          Web             Web              Web   
first_device_type           Mac Desktop     Mac Desktop  Windows Desktop   
first_browser                    Chrome          Chrome               IE   
country_destination                 NDF             NDF               US   

                                      3               4  
id                           bjjt8pjhuk      87mebub9p4  
date_account_created         2011-12-05      2010-09-14  
timestamp_first_active   20091031060129  20091208061105  
date_first_booking           2012-09-08      2010-02-18  
gender                           FEMALE       -unknown-  
age                                  42              41  
signup_method                  facebook           basic  
signup_flow                           0               0  
language                             en              en  
affiliate_channel                direct          direct  
affiliate_provider               direct          direct  
first_affiliate_tracked       untracked       untracked  
signup_app                          Web             Web  
first_device_type           Mac Desktop     Mac Desktop  
first_browser                   Firefox          Chrome  
country_destination               other              US

## Looking at sample data what can be observed 
???

   

### Looking at the structure of the Traveler data

 ### 1. Country Destination Values

- **country_destination** attribute (Most important) 

- Why? - Looking at the number of records that fall into each category can help provide some insights into how the model should be constructed as well as pitfalls to avoid.

<img src="./images/User_by_Destination.png" height="400" width="500"/>

#### Exercise 1: HW
  - Create the above table using Matplotlib or Seaborn

In [ ]:
### Start code




### End code

#### Analysis:???


#### What’s more, breaking down these percentage splits by year reveals that the percentage of users yet to make a booking (i.e., NDF) increases each year and reached over 60% in 2014.

<img src="./images/User_by_Destination_and_Year.png" height="400" width="500"/>


#### Exercise 2: HW
  - Create the above table using Matplotlib or Seaborn

In [ ]:
### Start code




### End code

### Analysis for building a learning model:

  ???
    
   
**Important:** Ensure that the training data has enough information to build a learning model that will predict other categories as well.

### 2. Account Creation Dates
 - **date_account_created** attribute – how values have changed over time?
 
<img src="./images/Accounts_Created_Over_Time.png" height="400" width="500"/>


#### Exercise 3: HW
  - Create the above line graph using Matplotlib or Seaborn

In [ ]:
### Start code




### End code

### Analysis:
  ???
  
   **Note:** Looking back to **sessions.csv** we would be limited to data from Jan 2014 onwards


### 3. Age Breakdown

#### Data Quality issue: 
- some users reported their ages well over 100 and 1000 


<img src="./images/Reported_Ages_of_Users.png" height="400" width="500"/>


#### Exercise 4: HW
  - Create the above bar graph using Matplotlib or Seaborn

In [ ]:
### Start code




### End code

### Analysis:
   ???
   
   <img src="./images/Missing_Ages.png" height="400" width="500"/>
   
   **Note:** While cleaning the data, need to decide what to do with these missing values.
        

### 4. First Device Type
 - **first_device_used** attribute
 
<img src="./images/First_Device_Used.png" height="400" width="500"/>



#### Exercise 5: HW
  - Create the above table using Matplotlib or Seaborn

In [ ]:
### Start code




### End code

### Analysis:
???

#### We can notice that the recent data is likely to be most useful for building the learning model.

#### Exercise 6: HW
- **Other attributes:** Give a look on other attributes and see how they can also help in building an accurate classification learning model.


In [ ]:
### Start code




### End code

### Milestone 2:  Focus on Cleaning Data

In [7]:
train_users.head().transpose()

0               1                2  \
id                           gxn3p5htnn      820tgsjxq7       4ft3gnwmtx   
date_account_created         2010-06-28      2011-05-25       2010-09-28   
timestamp_first_active   20090319043255  20090523174809   20090609231247   
date_first_booking                  NaN             NaN       2010-08-02   
gender                        -unknown-            MALE           FEMALE   
age                                 NaN              38               56   
signup_method                  facebook        facebook            basic   
signup_flow                           0               0                3   
language                             en              en               en   
affiliate_channel                direct             seo           direct   
affiliate_provider               direct          google           direct   
first_affiliate_tracked       untracked       untracked        untracked   
signup_app                          Web             Web              Web   
first_device_type           Mac Desktop     Mac Desktop  Windows Desktop   
first_browser                    Chrome          Chrome               IE   
country_destination                 NDF             NDF               US   

                                      3               4  
id                           bjjt8pjhuk      87mebub9p4  
date_account_created         2011-12-05      2010-09-14  
timestamp_first_active   20091031060129  20091208061105  
date_first_booking           2012-09-08      2010-02-18  
gender                           FEMALE       -unknown-  
age                                  42              41  
signup_method                  facebook           basic  
signup_flow                           0               0  
language                             en              en  
affiliate_channel                direct          direct  
affiliate_provider               direct          direct  
first_affiliate_tracked       untracked       untracked  
signup_app                          Web             Web  
first_device_type           Mac Desktop     Mac Desktop  
first_browser                   Firefox          Chrome  
country_destination               other              US

### [1] Fixing up timestamp formats 

  - **timestamp_first_active** attribute contains number like 20090609231247 instead of timestamps in the expected format: 2009-06-09 23:12:47
  
### [2] Filling in missing values 

### [3] Correcting erroneous values  

  - **gender** attribute where some have entered '-unknown-'. 
  - **age** attribute where some have entered value well over 100. 

### [4] Standardizing categories 
  
  - spelling mistakes, language differences or other factors will result in a given answer being provided in multiple ways.
  - **Example:** data on country of birth, if users are not provided with a standardized list of countries, the data will inevitably contain multiple spellings of the same country (e.g. USA, United States, U.S. and so on)


### Dealing with Missing Data - Solution for [2]

#### [1] Deleting/Ignoring rows with missing values
  - a. If more than 10% of data to be deleted, then try reconsidering.
  - b. Be confident that the rows being deleted do not contain information which is not contained in other rows.
       - For example, in the dataset we observe that many users have not provided their age. 
       - Can we assume that the people who chose not to provide their age are the same as the users who did? 
       -  Or are they likely to represent a different type of user, perhaps an older and more privacy conscious user, and therefore a user who is likely to make different choices on which countries to visit? 
       - If the answer is the latter, we probably do not want to just delete the records.


#### [2] Filling in the missing values
- What value to use?

- Depends on a range of factors, including the type of data we are trying to fill.

- **Categorical:** 
    - If the data is categorical (i.e. countries, device types, etc.), it may make sense to simply create a new category that will represent ‘unknown’.

    - Another option may be to fill the values with the most common value for that attribute (use mode).

    - Since these are broad methods for filling the missing values, they may **oversimplify** your data and/or make your final learning model less accurate.

- **Numerical:** 
    - For example age attribute, we could use mean or median to fill values.
    
    - Or, take an average based on some other criteria – for example filling the missing age values based on an average age for users who selected the same country_destination.

**Note:** For both types of data, we can use far more complicated methods to impute the missing values. There are endless ways...


#### Exercise 7: HW
- **Impute the missing values:** Explore and list the endless ways of filling missing values for boh Numerical and Categorical data largely practiced in competions for data cleaning.

In [ ]:
### List solutions:




### Cleaning efforts on two files

 - **train_users_2.csv** and 
 - **test_users.csv** 

In [22]:
#Loading the data

print("Reading data...")
train_file = "./traveler_dataset/train_users_2.csv"
df_train = pd.read_csv(train_file, header = 0,index_col = None)

test_file = "./traveler_dataset/test_users.csv"
df_test = pd.read_csv(test_file, header = 0,index_col = None)

# Combining into one dataset for cleaning
df_all = pd.concat((df_train, df_test), axis = 0, ignore_index = True, sort = False)
print("Reading data...completed")


Reading data...
Reading data...completed


#### Exercise 8:
- Do few operations like displaying the attributes, sample rows, finding NaNs, attribute statistics, no. of rows, columns, etc in dataframes 
  - **df_train**
  - **df_test**
  - **df_all**



In [9]:
### Start code




### End code

### Cleaning the timestamps - Fixing up formats of dates

- Why to convert? 
   - Reason: e.g. subtracting one date from another, extracting the month of the year from each date, etc.

**Note:** In next exercise, we will find its importance when we start adding various new features to the training data based on this date information.

In [23]:
# Fixing date formats in Pandas using to_datetime
## Change dates to specific format

print("Fixing timestamps...")
df_all['date_account_created'] = pd.to_datetime(df_all['date_account_created'], format = '%Y-%m-%d')
df_all['timestamp_first_active'] = pd.to_datetime(df_all['timestamp_first_active'], format = '%Y%m%d%H%M%S')
print("Fixing timestamps...completed")
df_all.head()


Fixing timestamps...
Fixing timestamps...completed


id date_account_created timestamp_first_active date_first_booking  \
0  gxn3p5htnn           2010-06-28    2009-03-19 04:32:55                NaN   
1  820tgsjxq7           2011-05-25    2009-05-23 17:48:09                NaN   
2  4ft3gnwmtx           2010-09-28    2009-06-09 23:12:47         2010-08-02   
3  bjjt8pjhuk           2011-12-05    2009-10-31 06:01:29         2012-09-08   
4  87mebub9p4           2010-09-14    2009-12-08 06:11:05         2010-02-18   

      gender   age signup_method  signup_flow language affiliate_channel  \
0  -unknown-   NaN      facebook            0       en            direct   
1       MALE  38.0      facebook            0       en               seo   
2     FEMALE  56.0         basic            3       en            direct   
3     FEMALE  42.0      facebook            0       en            direct   
4  -unknown-  41.0         basic            0       en            direct   

  affiliate_provider first_affiliate_tracked signup_app first_device_type  \
0             direct               untracked        Web       Mac Desktop   
1             google               untracked        Web       Mac Desktop   
2             direct               untracked        Web   Windows Desktop   
3             direct               untracked        Web       Mac Desktop   
4             direct               untracked        Web       Mac Desktop   

  first_browser country_destination  
0        Chrome                 NDF  
1        Chrome                 NDF  
2            IE                  US  
3       Firefox               other  
4        Chrome                  US

In [15]:
## Do following step only if date_account_created is empty
## date_account_created attribute is sometimes empty

## Solution: replace the empty values with the value in the timestamp_first_active attribute using the fillna() function
# df_all['date_account_created'].fillna(df_all.timestamp_first_active, inplace = True)

In [24]:
df_train.head().transpose()

0               1                2  \
id                           gxn3p5htnn      820tgsjxq7       4ft3gnwmtx   
date_account_created         2010-06-28      2011-05-25       2010-09-28   
timestamp_first_active   20090319043255  20090523174809   20090609231247   
date_first_booking                  NaN             NaN       2010-08-02   
gender                        -unknown-            MALE           FEMALE   
age                                 NaN              38               56   
signup_method                  facebook        facebook            basic   
signup_flow                           0               0                3   
language                             en              en               en   
affiliate_channel                direct             seo           direct   
affiliate_provider               direct          google           direct   
first_affiliate_tracked       untracked       untracked        untracked   
signup_app                          Web             Web              Web   
first_device_type           Mac Desktop     Mac Desktop  Windows Desktop   
first_browser                    Chrome          Chrome               IE   
country_destination                 NDF             NDF               US   

                                      3               4  
id                           bjjt8pjhuk      87mebub9p4  
date_account_created         2011-12-05      2010-09-14  
timestamp_first_active   20091031060129  20091208061105  
date_first_booking           2012-09-08      2010-02-18  
gender                           FEMALE       -unknown-  
age                                  42              41  
signup_method                  facebook           basic  
signup_flow                           0               0  
language                             en              en  
affiliate_channel                direct          direct  
affiliate_provider               direct          direct  
first_affiliate_tracked       untracked       untracked  
signup_app                          Web             Web  
first_device_type           Mac Desktop     Mac Desktop  
first_browser                   Firefox          Chrome  
country_destination               other              US

### Removing booking date field

- Why? Notice howmany date fields are there?

- We converted two date fields to a format above.

- Which one is not covered?

- Why? 
  - **Reason1:** In **training_users_2.csv**, all the users who have a first booking **country_destination** have a value in the **date_first_booking** attribute AND those who have not made a booking (i.e., **country_destination = NDF**) the value is missing. 
  
  - **Reason2:** In **test_users.csv**, the **date_first_booking** attribute is empty for all the records.


#### Analysis:

  ???

In [25]:
df_test.head().transpose()

0               1                2  \
id                           5uwns89zht      jtl0dijy2j       xx0ulgorjt   
date_account_created         2014-07-01      2014-07-01       2014-07-01   
timestamp_first_active   20140701000006  20140701000051   20140701000148   
date_first_booking                  NaN             NaN              NaN   
gender                           FEMALE       -unknown-        -unknown-   
age                                  35             NaN              NaN   
signup_method                  facebook           basic            basic   
signup_flow                           0               0                0   
language                             en              en               en   
affiliate_channel                direct          direct           direct   
affiliate_provider               direct          direct           direct   
first_affiliate_tracked       untracked       untracked           linked   
signup_app                        Moweb           Moweb              Web   
first_device_type                iPhone          iPhone  Windows Desktop   
first_browser             Mobile Safari   Mobile Safari           Chrome   

                                       3               4  
id                            6c6puo6ix0      czqhjk3yfe  
date_account_created          2014-07-01      2014-07-01  
timestamp_first_active    20140701000215  20140701000305  
date_first_booking                   NaN             NaN  
gender                         -unknown-       -unknown-  
age                                  NaN             NaN  
signup_method                      basic           basic  
signup_flow                            0               0  
language                              en              en  
affiliate_channel                 direct          direct  
affiliate_provider                direct          direct  
first_affiliate_tracked           linked       untracked  
signup_app                           Web             Web  
first_device_type        Windows Desktop     Mac Desktop  
first_browser                         IE          Safari

In [26]:
## Removing date_first_booking attribute
df_all.drop('date_first_booking', axis = 1, inplace = True)
print("Droped date_first_booking attribute...")

Droped date_first_booking attribute...


In [27]:
df_all.head()

id date_account_created timestamp_first_active     gender   age  \
0  gxn3p5htnn           2010-06-28    2009-03-19 04:32:55  -unknown-   NaN   
1  820tgsjxq7           2011-05-25    2009-05-23 17:48:09       MALE  38.0   
2  4ft3gnwmtx           2010-09-28    2009-06-09 23:12:47     FEMALE  56.0   
3  bjjt8pjhuk           2011-12-05    2009-10-31 06:01:29     FEMALE  42.0   
4  87mebub9p4           2010-09-14    2009-12-08 06:11:05  -unknown-  41.0   

  signup_method  signup_flow language affiliate_channel affiliate_provider  \
0      facebook            0       en            direct             direct   
1      facebook            0       en               seo             google   
2         basic            3       en            direct             direct   
3      facebook            0       en            direct             direct   
4         basic            0       en            direct             direct   

  first_affiliate_tracked signup_app first_device_type first_browser  \
0               untracked        Web       Mac Desktop        Chrome   
1               untracked        Web       Mac Desktop        Chrome   
2               untracked        Web   Windows Desktop            IE   
3               untracked        Web       Mac Desktop       Firefox   
4               untracked        Web       Mac Desktop        Chrome   

  country_destination  
0                 NDF  
1                 NDF  
2                  US  
3               other  
4                  US

### Correcting erroneous values - Solution for [3] (Cleaning the Age attribute)

- [1] **Outliers** - there are several age values that are clearly incorrect (unreasonably high or too low)
     - **Solution:** replace these incorrect values with 'NaN' (changing incorrect values into missing values)


- [2] **Missing values** - there are a significant number of users who did not provide their age at all, they show up as NaN in the dataset
     - **Solution:** change all the NaN values to -1

In [28]:
## Remove outliers function - [1]
def remove_outliers(df, column, min_val, max_val):
    col_values = df[column].values
    df[column] = np.where(np.logical_or(col_values <= min_val, col_values >= max_val), np.NaN, col_values)
    return df

print("Removing outliers in age attribute with NaN...")
df_all = remove_outliers(df = df_all, column = 'age', min_val = 15, max_val = 90)
print("Removing outliers in age attribute with NaN...completed")

## Fixing age column - [2]
print("Fixing age attribute...")
df_all['age'].fillna(-1, inplace = True)
print("Fixing age attribute...completed")


Removing outliers in age attribute with NaN...
Removing outliers in age attribute with NaN...completed
Fixing age attribute...
Fixing age attribute...completed


/home/shankar/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:4: RuntimeWarning: invalid value encountered in less_equal
  after removing the cwd from sys.path.
/home/shankar/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:4: RuntimeWarning: invalid value encountered in greater_equal
  after removing the cwd from sys.path.


#### Exercise 9: HW
- There are several more ways to fill in the missing values in the age attribute, try and list


In [18]:
### Start code



### End code

### Filling in missing values: Identify and fill additional attributes with missing values 

- On such attribute is **first_affiliate_tracked** has missing values
  - **Solution:** follow same procedure as above (change all the NaN values to -1)


In [29]:
df_all.head(10).transpose()

0                    1  \
id                                gxn3p5htnn           820tgsjxq7   
date_account_created     2010-06-28 00:00:00  2011-05-25 00:00:00   
timestamp_first_active   2009-03-19 04:32:55  2009-05-23 17:48:09   
gender                             -unknown-                 MALE   
age                                       -1                   38   
signup_method                       facebook             facebook   
signup_flow                                0                    0   
language                                  en                   en   
affiliate_channel                     direct                  seo   
affiliate_provider                    direct               google   
first_affiliate_tracked            untracked            untracked   
signup_app                               Web                  Web   
first_device_type                Mac Desktop          Mac Desktop   
first_browser                         Chrome               Chrome   
country_destination                      NDF                  NDF   

                                           2                    3  \
id                                4ft3gnwmtx           bjjt8pjhuk   
date_account_created     2010-09-28 00:00:00  2011-12-05 00:00:00   
timestamp_first_active   2009-06-09 23:12:47  2009-10-31 06:01:29   
gender                                FEMALE               FEMALE   
age                                       56                   42   
signup_method                          basic             facebook   
signup_flow                                3                    0   
language                                  en                   en   
affiliate_channel                     direct               direct   
affiliate_provider                    direct               direct   
first_affiliate_tracked            untracked            untracked   
signup_app                               Web                  Web   
first_device_type            Windows Desktop          Mac Desktop   
first_browser                             IE              Firefox   
country_destination                       US                other   

                                           4                    5  \
id                                87mebub9p4           osr2jwljor   
date_account_created     2010-09-14 00:00:00  2010-01-01 00:00:00   
timestamp_first_active   2009-12-08 06:11:05  2010-01-01 21:56:19   
gender                             -unknown-            -unknown-   
age                                       41                   -1   
signup_method                          basic                basic   
signup_flow                                0                    0   
language                                  en                   en   
affiliate_channel                     direct                other   
affiliate_provider                    direct                other   
first_affiliate_tracked            untracked                  omg   
signup_app                               Web                  Web   
first_device_type                Mac Desktop          Mac Desktop   
first_browser                         Chrome               Chrome   
country_destination                       US                   US   

                                           6                    7  \
id                                lsw9q7uk0j           0d01nltbrs   
date_account_created     2010-01-02 00:00:00  2010-01-03 00:00:00   
timestamp_first_active   2010-01-02 01:25:58  2010-01-03 19:19:05   
gender                                FEMALE               FEMALE   
age                                       46                   47   
signup_method                          basic                basic   
signup_flow                                0                    0   
language                                  en                   en   
affiliate_channel                      other               direct   
affiliate_provider                crai

#### Exercise 10:
- Are there any missing values in **first_affiliate_tracked** attribute, if yes how many?

In [ ]:
### Start code



### End code

#### Exercise 11:
- Fill **first_affiliate_tracked** attribute by following same procedure as above (change all the NaN values to -1)

In [30]:
# Fill first_affiliate_tracked attribute
### Start code
print("Filling first_affiliate_tracked attribute...")

df_all['first_affiliate_tracked'].fillna(-1, inplace=True)

print("Filling first_affiliate_tracked attribute...completed")
### End code

Filling first_affiliate_tracked attribute...
Filling first_affiliate_tracked attribute...completed


### How does the data look like after all these changes? 

In [13]:
## Sample of some rows from cleaned dataset
df_all.head(10)

id date_account_created timestamp_first_active     gender   age  \
0  gxn3p5htnn           2010-06-28    2009-03-19 04:32:55  -unknown-  -1.0   
1  820tgsjxq7           2011-05-25    2009-05-23 17:48:09       MALE  38.0   
2  4ft3gnwmtx           2010-09-28    2009-06-09 23:12:47     FEMALE  56.0   
3  bjjt8pjhuk           2011-12-05    2009-10-31 06:01:29     FEMALE  42.0   
4  87mebub9p4           2010-09-14    2009-12-08 06:11:05  -unknown-  41.0   
5  osr2jwljor           2010-01-01    2010-01-01 21:56:19  -unknown-  -1.0   
6  lsw9q7uk0j           2010-01-02    2010-01-02 01:25:58     FEMALE  46.0   
7  0d01nltbrs           2010-01-03    2010-01-03 19:19:05     FEMALE  47.0   
8  a1vcnhxeij           2010-01-04    2010-01-04 00:42:11     FEMALE  50.0   
9  6uh8zyj2gn           2010-01-04    2010-01-04 02:37:58  -unknown-  46.0   

  signup_method  signup_flow language affiliate_channel affiliate_provider  \
0      facebook            0       en            direct             direct   
1      facebook            0       en               seo             google   
2         basic            3       en            direct             direct   
3      facebook            0       en            direct             direct   
4         basic            0       en            direct             direct   
5         basic            0       en             other              other   
6         basic            0       en             other         craigslist   
7         basic            0       en            direct             direct   
8         basic            0       en             other         craigslist   
9         basic            0       en             other         craigslist   

  first_affiliate_tracked signup_app first_device_type first_browser  \
0               untracked        Web       Mac Desktop        Chrome   
1               untracked        Web       Mac Desktop        Chrome   
2               untracked        Web   Windows Desktop            IE   
3               untracked        Web       Mac Desktop       Firefox   
4               untracked        Web       Mac Desktop        Chrome   
5                     omg        Web       Mac Desktop        Chrome   
6               untracked        Web       Mac Desktop        Safari   
7                     omg        Web       Mac Desktop        Safari   
8               untracked        Web       Mac Desktop        Safari   
9                     omg        Web       Mac Desktop       Firefox   

  country_destination  
0                 NDF  
1                 NDF  
2                  US  
3               other  
4                  US  
5                  US  
6                  US  
7                  US  
8                  US  
9                  US

### Is that all?
- **Not really** - this is just a small work of cleaning, you need to try more

### What Next?
- **Aim:** Focus on transforming the data and feature extraction
   - **Why?** To build better prediction learning model.

## Data Transformation and Feature Extraction as a Concept

### Why transformation?

- Unlike the steps taken during cleaning, which are designed to address problems with the raw data (missing and erroneous values, formatting issues etc.), 

- Data Transformation steps change the values and/or structure of the data (data transformation) and add additional features (feature extraction).


### Data Transformation methods:

#### [1] Bucketing/Binning

- the numerical values in a particular attribute are converted from a continuous series into fixed ranges
- example, instead of using the age value of all users, we could place them into buckets such as 15-20 years old, 21-25 years old and so on

#### [2] Normalization

#### [3] Other Transformations

- There are unlimited number of ways that the numerical values of a given attribute can be transformed such that they are more suitable for the algorithm being used.

- Logarithm function: This transformation is a commonly used method of dealing with exponential data series (i.e. a column where there are lot of low values and relatively few high values). 

#### [3.1] One Hot Encoding (used for categorical data)
<img src="./images/One_Hot_Encoding.png" height="400" width="500"/>

### Feature Extraction:
- feature construction and feature selection to add to dataset

#### [1] Using Hierarchical Information

 - data in dataset represents one level of hierarchy, and extracting other implied levels of that hierarchy will provide the learning model with useful information
 
 - Eg: Consider date fields,  
     - extracting the day of the week, 
     - the month of the year,
     - the hour of the day, 
     
 - could add important information for the algorithm to use
     - Maybe people who create their accounts in summer months are more likely to make a booking in a warmer country. 
     - Maybe people who were first active late at night are more disorganized travelers and are therefore more likely to make a domestic first booking. 
     - Combination of these factors may make the difference (e.g. users first active late at night, in the summer months, on a weekday are more likely to travel to Portugal). 
   
   #### Note:The point is not to be able to explain why a factor may be important, but to think of as many factors as possible to test, and allow the algorithm to determine what is important and not important.
    
#### [2] Adding External Data

 - (known as record linkage) data enrichment
 
 - Eg: Countries could be enriched with demographic data about the country such as 
      - population, 
      - income per capita or 
      - land area 
 - all the factors that may allow the algorithm to draw conclusions across similar groups of countries.
 
 - Consider how much more accurately we could predict a first booking country of a user if we could link the data from their TRAVELER profile to data from one of their social media profiles (Facebook, Twitter etc.) or even better, from other Traveler accounts.
 
 
 #### IMPORTANT:
 
 - The key point here is that it is worth investing time looking for ways to add new and useful data to your existing dataset before moving onto the learning modeling step. 
 
 - Expanding your dataset in this manner will often produce far bigger improvements in prediction accuracy than the choice of algorithm or the tuning of the algorithm parameters.

### Practical: Transforming Categorical Data
#### One Hot Enconder method 

   - replacing the categorical fields in the dataset with multiple columns representing one value from each column

   - with Scikit Learn (http://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.OneHotEncoder.html)
   
   - own function

In [31]:
## Prototype for only gender attribute  - testing
# User defined One Hot Encoding function
def convert_to_binary(df,column_to_convert):
    categories = list(df[column_to_convert].drop_duplicates())
    print (categories)
    for category in categories:
        cat_name = str(category).replace(" ", "_").replace("(", "").replace(")", "").replace("/", "_").replace("-", "").lower()
        col_name = column_to_convert[:5] + '_' + cat_name[:10]
        print (col_name)
        df[col_name] = 0
        df.loc[(df[column_to_convert] == category), col_name] = 1

    return df

# One Hot Encoding
print("One Hot Encoding categorical data...")
#columns_to_convert = ['gender', 'signup_method', 'signup_flow', 'language', 'affiliate_channel', 'affiliate_provider', 'first_affiliate_tracked', 'signup_app', 'first_device_type', 'first_browser']
columns_to_convert = ['gender']
for column in columns_to_convert:
    df_all = convert_to_binary(df=df_all, column_to_convert=column)
    #df_all.drop(column, axis=1, inplace=True)
print("One Hot Encoding categorical data...completed")
df_all.head()

One Hot Encoding categorical data...
['-unknown-', 'MALE', 'FEMALE', 'OTHER']
gende_unknown
gende_male
gende_female
gende_other
One Hot Encoding categorical data...completed


id date_account_created timestamp_first_active     gender   age  \
0  gxn3p5htnn           2010-06-28    2009-03-19 04:32:55  -unknown-  -1.0   
1  820tgsjxq7           2011-05-25    2009-05-23 17:48:09       MALE  38.0   
2  4ft3gnwmtx           2010-09-28    2009-06-09 23:12:47     FEMALE  56.0   
3  bjjt8pjhuk           2011-12-05    2009-10-31 06:01:29     FEMALE  42.0   
4  87mebub9p4           2010-09-14    2009-12-08 06:11:05  -unknown-  41.0   

  signup_method  signup_flow language affiliate_channel affiliate_provider  \
0      facebook            0       en            direct             direct   
1      facebook            0       en               seo             google   
2         basic            3       en            direct             direct   
3      facebook            0       en            direct             direct   
4         basic            0       en            direct             direct   

  first_affiliate_tracked signup_app first_device_type first_browser  \
0               untracked        Web       Mac Desktop        Chrome   
1               untracked        Web       Mac Desktop        Chrome   
2               untracked        Web   Windows Desktop            IE   
3               untracked        Web       Mac Desktop       Firefox   
4               untracked        Web       Mac Desktop        Chrome   

  country_destination  gende_unknown  gende_male  gende_female  gende_other  
0                 NDF              1           0             0            0  
1                 NDF              0           1             0            0  
2                  US              0           0             1            0  
3               other              0           0             1            0  
4                  US              1           0             0            0

In [32]:
# User defined One Hot Encoding function
def convert_to_binary(df, column_to_convert):
    categories = list(df[column_to_convert].drop_duplicates())

    for category in categories:
        cat_name = str(category).replace(" ", "_").replace("(", "").replace(")", "").replace("/", "_").replace("-", "").lower()
        col_name = column_to_convert[:5] + '_' + cat_name[:10]
        df[col_name] = 0
        df.loc[(df[column_to_convert] == category), col_name] = 1

    return df

# One Hot Encoding
print("One Hot Encoding categorical data...")
columns_to_convert = ['gender', 'signup_method', 'signup_flow', 'language', 'affiliate_channel', 'affiliate_provider', 'first_affiliate_tracked', 'signup_app', 'first_device_type', 'first_browser']

for column in columns_to_convert:
    df_all = convert_to_binary(df=df_all, column_to_convert=column)
    df_all.drop(column, axis=1, inplace=True)
print("One Hot Encoding categorical data...completed")

One Hot Encoding categorical data...
One Hot Encoding categorical data...completed


In [33]:
df_all.head()

id date_account_created timestamp_first_active   age  \
0  gxn3p5htnn           2010-06-28    2009-03-19 04:32:55  -1.0   
1  820tgsjxq7           2011-05-25    2009-05-23 17:48:09  38.0   
2  4ft3gnwmtx           2010-09-28    2009-06-09 23:12:47  56.0   
3  bjjt8pjhuk           2011-12-05    2009-10-31 06:01:29  42.0   
4  87mebub9p4           2010-09-14    2009-12-08 06:11:05  41.0   

  country_destination  gende_unknown  gende_male  gende_female  gende_other  \
0                 NDF              1           0             0            0   
1                 NDF              0           1             0            0   
2                  US              0           0             1            0   
3               other              0           0             1            0   
4                  US              1           0             0            0   

   signu_facebook        ...         first_theworld_b  first_slimbrowse  \
0               1        ...                        0                 0   
1               1        ...                        0                 0   
2               0        ...                        0                 0   
3               1        ...                        0                 0   
4               0        ...                        0                 0   

   first_epic  first_stainless  first_googlebot  first_outlook_20  \
0           0                0                0                 0   
1           0                0                0                 0   
2           0                0                0                 0   
3           0                0                0                 0   
4           0                0                0                 0   

   first_icedragon  first_ibrowse  first_nintendo_b  first_uc_browser  
0                0              0                 0                 0  
1                0              0                 0                 0  
2                0              0                 0                 0  
3                0              0                 0                 0  
4                0              0                 0                 0  

[5 rows x 157 columns]

#### Creating New Features

- Using two date fields – **date_account_created** and **timestamp_first_active**

- Extract all the information we can out of these two date fields that could potentially differentiate which country someone will make their first booking in. 

In [34]:
# Add new date related fields
print("Adding new fields...")
df_all['day_account_created'] = df_all['date_account_created'].dt.weekday
df_all['month_account_created'] = df_all['date_account_created'].dt.month
df_all['quarter_account_created'] = df_all['date_account_created'].dt.quarter
df_all['year_account_created'] = df_all['date_account_created'].dt.year
df_all['hour_first_active'] = df_all['timestamp_first_active'].dt.hour
df_all['day_first_active'] = df_all['timestamp_first_active'].dt.weekday
df_all['month_first_active'] = df_all['timestamp_first_active'].dt.month
df_all['quarter_first_active'] = df_all['timestamp_first_active'].dt.quarter
df_all['year_first_active'] = df_all['timestamp_first_active'].dt.year
df_all['created_less_active'] = (df_all['date_account_created'] - df_all['timestamp_first_active']).dt.days
print("Adding new fields...completed")

# Drop unnecessary columns
print("Droping fields...")
columns_to_drop = ['date_account_created', 'timestamp_first_active', 'date_first_booking', 'country_destination']
for column in columns_to_drop:
    if column in df_all.columns:
        df_all.drop(column, axis=1, inplace=True)
print("Droping fields...completed")

Adding new fields...
Adding new fields...completed
Droping fields...
Droping fields...completed


In [35]:
df_all.head()

id   age  gende_unknown  gende_male  gende_female  gende_other  \
0  gxn3p5htnn  -1.0              1           0             0            0   
1  820tgsjxq7  38.0              0           1             0            0   
2  4ft3gnwmtx  56.0              0           0             1            0   
3  bjjt8pjhuk  42.0              0           0             1            0   
4  87mebub9p4  41.0              1           0             0            0   

   signu_facebook  signu_basic  signu_google  signu_weibo  \
0               1            0             0            0   
1               1            0             0            0   
2               0            1             0            0   
3               1            0             0            0   
4               0            1             0            0   

          ...           day_account_created  month_account_created  \
0         ...                             0                      6   
1         ...                             2                      5   
2         ...                             1                      9   
3         ...                             0                     12   
4         ...                             1                      9   

   quarter_account_created  year_account_created  hour_first_active  \
0                        2                  2010                  4   
1                        2                  2011                 17   
2                        3                  2010                 23   
3                        4                  2011                  6   
4                        3                  2010                  6   

   day_first_active  month_first_active  quarter_first_active  \
0                 3                   3                     1   
1                 5                   5                     2   
2                 1                   6                     2   
3                 5                  10                     4   
4                 1                  12                     4   

   year_first_active  created_less_active  
0               2009                  465  
1               2009                  731  
2               2009                  475  
3               2009                  764  
4               2009                  279  

[5 rows x 164 columns]

#### Summary

- Changed training dataset from 15 columns to 164 columns
- **HW:** Investigate what information could be extracted from the other non-date columns?
- **HW:** Check what external data could be added? 

- **HW:** With dummy variables created by One-hot-encoding are we introducing multicollinearity ?
    - If yes... can we use PCA to reduce the dimensionality down to prevent the model from being misled?
    
- **HW:** How worried about multicollinearity you need to be if you are (i) building a regression model and (ii) concerned with prediction accuracy ?

- **Example:** StackExchange multicollinearity: https://stats.stackexchange.com/questions/168622/why-is-multicollinearity-not-checked-in-modern-statistics-machine-learning